File: core/src/main/scala/nclogic/model/expr/Neg.scala
package nclogic.model.expr

case class Neg(e: Expr) extends Expr {
  override def toString = "!" + e

  def isAtomic = e match {
    case Next(_) => false
    case Globally(_) => false
    case Finally(_) => false
    case Neg(_) => false
    case True => false
    case False => false
    case _ => e.isAtomic
  }

  def simplify(implicit level: Int): Expr =
    if (isAtomic) this
    else e match {
      case True => False
      case False => True
      case Neg(x) => x.simplify
      case And(es) => Or(es.map(_.not)).simplify
      case Or(es) => And(es.map(_.not)).simplify
      case Next(x) => Next(Neg(x)).simplify
      case Globally(x) => Finally(Neg(x)).simplify
      case Finally(x) => Globally(Neg(x)).simplify
      case _ => Neg(e.simplify).simplify
    }

  override def baseTerms: Set[Expr] = e.baseTerms

  override def level: Int = e.level
}

File: javaapi/src/main/java/nclogic/java/model/expr/Neg.java
package nclogic.java.model.expr;

public class Neg extends Expr {

    private final nclogic.model.expr.Neg neg;
    private final Expr e;

    public Neg(Expr e) {
        this.e = e;
        neg = new nclogic.model.expr.Neg(e.getScalaExpr());
    }

    public Neg(nclogic.model.expr.Neg neg) {
        this.neg = neg;
        this.e = ToJavaExprConverter.convert(neg.e());
    }

    public Expr getE() {
        return e;
    }

    @Override
    public nclogic.model.expr.Expr getScalaExpr() {
        return neg;
    }
}
